File: conformance/cpp/events.cpp
// Conformance consumer: events sample, C++ target (ABI revision 2).
//
// Drives the generated header-only wrapper over the callback-interface and
// reference-counted-object surface:
//  - `Subscriber` is an abstract class the consumer subclasses; a
//    `std::shared_ptr<Subscriber>` crosses into the producer, which calls
//    `route`, `on_message`, and `on_attached` through a static vtable and
//    releases the box (running the consumer's destructor) when it drops its
//    last reference.
//  - `EventBus` is a copyable RAII wrapper: copies `_clone` the producer
//    object and share it; moves transfer the handle; destructors release.
//  - the `bus` handed to `on_attached` is an owned wrapper the consumer may
//    keep and use.
//  - `Delivery` return values steer `publish`'s accepted count.
//  - a subscriber that throws surfaces to the caller as `WeaveFFIError`
//    with code -4 and leaves the bus usable.
//  - `messages()` is a lazy single-pass range; `last_message()` is an
//    optional record; `publish_later` is a std::future settled from a
//    producer thread.
// Exits non-zero on the first failed check.

#include <cstdio>
#include <cstdlib>
#include <future>
#include <memory>
#include <optional>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

#include "weaveffi.hpp"

using namespace weaveffi;

static void check(bool ok, const char* what) {
    if (!ok) {
        std::fprintf(stderr, "cpp/events: FAIL: %s\n", what);
        std::exit(1);
    }
}

// State outlives the subscriber so the test can inspect it after the
// producer has freed the implementation.
struct SubState {
    std::vector<std::string> routed;
    std::vector<Message> received;
    int attached = 0;
    int64_t attached_count_seen = -1;
    const weaveffi_events_EventBus* attached_handle = nullptr;
    std::optional<EventBus> kept_bus;
    int freed = 0;
};

class RecordingSubscriber : public Subscriber {
    std::shared_ptr<SubState> state_;
    std::string skip_topic_;
    std::string stop_topic_;
    bool keep_bus_;

public:
    RecordingSubscriber(std::shared_ptr<SubState> state, std::string skip_topic,
                        std::string stop_topic, bool keep_bus)
        : state_(std::move(state)),
          skip_topic_(std::move(skip_topic)),
          stop_topic_(std::move(stop_topic)),
          keep_bus_(keep_bus) {}

    ~RecordingSubscriber() override { state_->freed++; }

    Delivery route(const std::string& topic) override {
        state_->routed.push_back(topic);
        if (topic == skip_topic_) return Delivery::Skip;
        if (topic == stop_topic_) return Delivery::AcceptAndStop;
        return Delivery::Accept;
    }

    int64_t on_message(const Message& message) override {
        state_->received.push_back(message);
        return static_cast<int64_t>(state_->received.size());
    }

    void on_attached(EventBus bus) override {
        state_->attached++;
        state_->attached_handle = bus.handle();
        // The wrapper is fully usable inside the callback; `subscribe` calls
        // `on_attached` before it pushes the subscriber, so the count is the
        // pre-subscribe value.
        state_->attached_count_seen = bus.subscriber_count();
        if (keep_bus_) state_->kept_bus = std::move(bus);
        // Otherwise `bus` goes out of scope here and releases its reference.
    }
};

// Throws from `route` for one topic and from `on_message` for one text.
class ThrowingSubscriber : public Subscriber {
    std::string route_bomb_;
    std::string message_bomb_;

public:
    ThrowingSubscriber(std::string route_bomb, std::string message_bomb)
        : route_bomb_(std::move(route_bomb)), message_bomb_(std::move(message_bomb)) {}

    Delivery route(const std::string& topic) override {
        if (topic == route_bomb_) throw std::runtime_error("route rejected " + topic);
        return Delivery::Accept;
    }

    int64_t on_message(const Message& message) override {
        if (message.text == message_bomb_) throw std::logic_error("on_message exploded");
        return 1;
    }

    void on_attached(EventBus) override {}
};

int main() {
    check_abi_version();

    EventBus bus;
    check(bus.handle() != nullptr, "constructor yields a live handle");
    check(bus.subscriber_count() == 0, "fresh bus has no subscribers");
    check(!bus.last_message().has_value(), "fresh bus has no last message");
    {
        int n = 0;
        for (auto&& m : bus.messages()) {
            (void)m;
            n++;
        }
        check(n == 0, "fresh bus streams no messages");
    }

    // Copy semantics: a copy shares the object (same pointer), both usable.
    {
        EventBus copy = bus;
        check(copy.handle() == bus.handle(), "copy constructor clones the same object");
        check(copy.subscriber_count() == 0, "copy is usable");
        EventBus assigned;
        check(assigned.handle() != bus.handle(), "a second constructor makes a distinct object");
        assigned = bus;
        check(assigned.handle() == bus.handle(), "copy assignment clones the same object");
        EventBus& same = assigned;
        assigned = same;
        check(assigned.handle() == bus.handle(), "self-assignment is a no-op");
        EventBus moved = std::move(copy);
        check(moved.handle() == bus.handle(), "move constructor transfers the handle");
        check(copy.handle() == nullptr, "moved-from wrapper is empty");
        // All three destruct here: two releases plus a harmless empty one.
    }
    check(bus.subscriber_count() == 0, "bus survives its copies' destructors");

    // Subscribers: A skips "quiet", B stops on "stop", C accepts everything.
    auto a_state = std::make_shared<SubState>();
    auto b_state = std::make_shared<SubState>();
    auto c_state = std::make_shared<SubState>();
    check(bus.subscribe(std::make_shared<RecordingSubscriber>(a_state, "quiet", "", true)) == 1,
          "subscribe returns 1");
    check(a_state->attached == 1, "on_attached fired once for A");
    check(a_state->attached_handle == bus.handle(), "on_attached received the same bus object");
    check(a_state->attached_count_seen == 0, "on_attached ran before the subscriber was retained");
    check(a_state->kept_bus.has_value(), "A kept the bus handed to it");
    check(a_state->kept_bus->subscriber_count() == 1, "the kept bus is usable after the callback");

    check(bus.subscribe(std::make_shared<RecordingSubscriber>(b_state, "", "stop", false)) == 2,
          "subscribe returns 2");
    check(b_state->attached == 1 && b_state->attached_count_seen == 1, "B attached second");
    check(bus.subscribe(std::make_shared<RecordingSubscriber>(c_state, "", "", false)) == 3,
          "subscribe returns 3");
    check(bus.subscriber_count() == 3, "subscriber_count is 3");
    check(a_state->freed == 0 && b_state->freed == 0 && c_state->freed == 0,
          "retained subscribers are not freed");

    // Everyone accepts.
    check(bus.publish("news", "hello", {"a", "b"}) == 3, "publish delivers to all three");
    check(a_state->received.size() == 1 && b_state->received.size() == 1 &&
              c_state->received.size() == 1,
          "each subscriber received one message");
    {
        const Message& m = a_state->received[0];
        check(m.seq == 1, "first message has seq 1");
        check(m.topic == "news", "message topic");
        check(m.text == "hello", "message text");
        check(m.tags.size() == 2 && m.tags[0] == "a" && m.tags[1] == "b", "message tags");
    }
    check(a_state->routed.size() == 1 && a_state->routed[0] == "news", "route saw the topic");

    // A skips "quiet".
    check(bus.publish("quiet", "psst", {}) == 2, "Skip lowers the accepted count");
    check(a_state->received.size() == 1, "A was skipped");
    check(a_state->routed.size() == 2 && a_state->routed[1] == "quiet", "A was still asked");
    check(b_state->received.size() == 2 && b_state->received[1].seq == 2, "B got seq 2");
    check(c_state->received[1].tags.empty(), "empty tags list round-trips");

    // B stops the chain, so C is neither asked nor delivered.
    check(bus.publish("stop", "last", {"x"}) == 2, "AcceptAndStop halts later subscribers");
    check(a_state->received.size() == 2, "A accepted before the stop");
    check(b_state->received.size() == 3, "B accepted with stop");
    check(c_state->received.size() == 2, "C did not receive after the stop");
    check(c_state->routed.size() == 2, "C was not asked after the stop");

    // Iterator: one producer `next` per step, in publish order.
    {
        std::vector<std::string> texts;
        for (auto&& t : bus.messages()) texts.push_back(t);
        check(texts.size() == 3, "messages streams three texts");
        check(texts[0] == "hello" && texts[1] == "psst" && texts[2] == "last",
              "messages are in publish order");
    }
    // Abandoning the range early destroys the producer iterator via RAII.
    {
        auto range = bus.messages();
        auto it = range.begin();
        check(*it == "hello", "early-abandoned range yields the first element");
    }
    // Manual `next()` drains and then reports exhaustion.
    {
        auto range = bus.messages();
        int n = 0;
        while (range.next().has_value()) n++;
        check(n == 3, "manual next() yields three elements");
        check(!range.next().has_value(), "exhausted range stays exhausted");
    }

    // Optional record return.
    {
        std::optional<Message> last = bus.last_message();
        check(last.has_value(), "last_message is present");
        check(last->seq == 3 && last->topic == "stop" && last->text == "last",
              "last_message is the third message");
        check(last->tags.size() == 1 && last->tags[0] == "x", "last_message tags");
    }

    // Async: settled from a producer thread; subscribers run there too.
    {
        std::future<int64_t> pending = bus.publish_later("later", "async");
        check(pending.get() == 3, "publish_later resolves with the accepted count");
        check(a_state->received.size() == 3 && a_state->received[2].seq == 4,
              "A received the async message");
        check(b_state->received.size() == 4 && b_state->received[3].text == "async",
              "B received the async message");
        check(c_state->received.size() == 3 && c_state->received[2].topic == "later",
              "C received the async message");
        std::optional<Message> last = bus.last_message();
        check(last.has_value() && last->seq == 4 && last->tags.empty(),
              "async publish is logged with empty tags");
    }

    // Free function taking a callback interface: routes without retaining.
    {
        auto tmp_state = std::make_shared<SubState>();
        auto tmp = std::make_shared<RecordingSubscriber>(tmp_state, "quiet", "stop", false);
        std::weak_ptr<Subscriber> weak = tmp;
        check(events::route_once(std::move(tmp), "quiet") == Delivery::Skip, "route_once Skip");
        check(weak.expired() && tmp_state->freed == 1,
              "route_once released the subscriber when the call returned");
        tmp_state = std::make_shared<SubState>();
        check(events::route_once(std::make_shared<RecordingSubscriber>(tmp_state, "", "stop", false),
                                 "stop") == Delivery::AcceptAndStop,
              "route_once AcceptAndStop");
        check(events::route_once(std::make_shared<RecordingSubscriber>(tmp_state, "", "", false),
                                 "other") == Delivery::Accept,
              "route_once Accept");
        check(tmp_state->attached == 0, "route_once never calls on_attached");
    }

    // A null callback interface is rejected before crossing the ABI.
    {
        bool threw = false;
        try {
            bus.subscribe(nullptr);
        } catch (const std::invalid_argument&) {
            threw = true;
        }
        check(threw, "null subscriber throws std::invalid_argument");
        check(bus.subscriber_count() == 3, "null subscribe left the bus unchanged");
    }

    // Foreign errors: a throwing implementation aborts the call with -4.
    {
        EventBus bus2;
        auto ok_state = std::make_shared<SubState>();
        check(bus2.subscribe(std::make_shared<RecordingSubscriber>(ok_state, "", "", false)) == 1,
              "bus2 first subscriber");
        check(bus2.subscribe(std::make_shared<ThrowingSubscriber>("boom", "explode")) == 2,
              "bus2 throwing subscriber");

        bool caught = false;
        try {
            bus2.publish("boom", "x", {});
        } catch (const WeaveFFIError& e) {
            caught = true;
            check(e.code() == -4, "route exception maps to FOREIGN_ERROR_CODE (-4)");
            check(std::string(e.what()).find("route rejected boom") != std::string::npos,
                  "foreign error carries the C++ exception message");
        }
        check(caught, "publish threw for a throwing route");
        check(ok_state->received.size() == 1, "earlier subscriber was delivered before the abort");

        caught = false;
        try {
            bus2.publish("fine", "explode", {});
        } catch (const WeaveFFIError& e) {
            caught = true;
            check(e.code() == -4, "on_message exception maps to -4");
            check(std::string(e.what()).find("on_message exploded") != std::string::npos,
                  "on_message foreign error message");
        }
        check(caught, "publish threw for a throwing on_message");

        // The bus is still usable afterward, and the log kept both messages.
        check(bus2.publish("ok", "y", {}) == 2, "bus is usable after a foreign error");
        std::optional<Message> last = bus2.last_message();
        check(last.has_value() && last->seq == 3, "aborted publishes were still logged");

        // The async path surfaces the same code through the future.
        caught = false;
        try {
            bus2.publish_later("boom", "z").get();
        } catch (const WeaveFFIError& e) {
            caught = true;
            check(e.code() == -4, "async foreign error maps to -4");
        }
        check(caught, "publish_later future rethrows the foreign error");

        // route_once with a throwing subscriber.
        caught = false;
        try {
            events::route_once(std::make_shared<ThrowingSubscriber>("boom", ""), "boom");
        } catch (const WeaveFFIError& e) {
            caught = (e.code() == -4);
        }
        check(caught, "route_once surfaces the foreign error");

        bus2.clear_subscribers();
        check(ok_state->freed == 1, "clearing bus2 freed its subscribers");
    }

    // Releasing: clear_subscribers drops the producer's references, which
    // runs every implementation's destructor (A's releases its kept bus).
    bus.clear_subscribers();
    check(bus.subscriber_count() == 0, "clear_subscribers empties the bus");
    check(a_state->freed == 1 && b_state->freed == 1 && c_state->freed == 1,
          "free ran exactly once per subscriber");
    // The wrapper A stashed in its state still owns a reference and is
    // usable; dropping it releases that reference.
    check(a_state->kept_bus.has_value() && a_state->kept_bus->handle() == bus.handle(),
          "kept bus still points at the bus");
    check(a_state->kept_bus->subscriber_count() == 0, "kept bus is usable after free");
    a_state->kept_bus.reset();
    check(bus.publish("after", "clear", {}) == 0, "publishing with no subscribers accepts none");
    check(bus.last_message()->seq == 5, "log kept counting");

    // `bus` releases the last reference when main returns.
    std::printf("cpp/events: OK\n");
    return 0;
}

File: conformance/swift/events.swift
// Conformance consumer: events sample, Swift target.
//
// Binds through the generated `Events` module and drives the ABI 2 surface:
// the `Subscriber` callback interface implemented as a Swift class (every
// method observed with its arguments, including the `EventBus` object the
// producer hands to `onAttached`, which is usable and released), `Delivery`
// return values steering `publish`'s accepted count, the reference-counted
// `EventBus` object, the `publishLater` async wrapper, `messages()` as a lazy
// Sequence, the `lastMessage()` optional record, and the free function
// `routeOnce`. Release is observed through weak references: a subscriber's
// box is freed when the bus drops it, and the bus wrapper handed to a callback
// is destroyed when the consumer lets go of it. The events module declares no
// error domain, so every wrapper is non-throwing and traps on a failure
// (including a subscriber that throws, which the producer reports as code
// -4); the throwing-callback path is exercised in the kvstore lane, whose
// wrappers do throw. Exits non-zero on any mismatch.

import Foundation
import Events

func fail(_ msg: String) -> Never {
    FileHandle.standardError.write(Data("assertion failed: \(msg)\n".utf8))
    exit(1)
}

func expect(_ cond: Bool, _ msg: String) {
    if !cond { fail(msg) }
}

/// A consumer-side subscriber. Records every callback so the test can assert
/// the arguments the producer passed, and counts live instances so the
/// producer's `free` (which releases the generated box, and with it the last
/// strong reference to this object) is observable.
final class TestSubscriber: Subscriber {
    static var live = 0

    let name: String
    let skipTopic: String
    let keepBus: Bool
    var routed: [String] = []
    var received: [Message] = []
    var attachedCount = 0
    /// Subscriber count observed through the bus object inside `onAttached`.
    var countSeenOnAttach: Int64 = -1
    /// The bus adopted from `onAttached` when `keepBus` is set.
    var attachedBus: EventBus?
    /// A weak view of the bus wrapper handed to `onAttached`, to prove the
    /// wrapper (and its strong reference) was released when not kept.
    weak var weakAttachedBus: EventBus?

    init(name: String, skipTopic: String = "", keepBus: Bool = false) {
        self.name = name
        self.skipTopic = skipTopic
        self.keepBus = keepBus
        TestSubscriber.live += 1
    }

    deinit { TestSubscriber.live -= 1 }

    func route(topic: String) throws -> Delivery {
        routed.append(topic)
        if topic == skipTopic { return .skip }
        if topic == "stop" { return .acceptAndStop }
        return .accept
    }

    func onMessage(message: Message) throws -> Int64 {
        received.append(message)
        return Int64(received.count)
    }

    func onAttached(bus: EventBus) throws {
        attachedCount += 1
        // The object is live and usable inside the callback: `subscribe`
        // attaches before it pushes, so the count excludes this subscriber.
        countSeenOnAttach = bus.subscriberCount()
        weakAttachedBus = bus
        if keepBus { attachedBus = bus }
    }
}

// --- Empty bus ---------------------------------------------------------------
var bus: EventBus? = EventBus()
expect(bus!.subscriberCount() == 0, "new bus has no subscribers")
expect(bus!.lastMessage() == nil, "new bus has no last message")
expect(Array(bus!.messages()).isEmpty, "new bus has no messages")

// --- subscribe: onAttached receives a usable bus object ----------------------
var a: TestSubscriber? = TestSubscriber(name: "a", skipTopic: "quiet")
var b: TestSubscriber? = TestSubscriber(name: "b", keepBus: true)
weak var weakA = a
weak var weakB = b

expect(bus!.subscribe(subscriber: a!) == 1, "first subscribe returns 1")
expect(a!.attachedCount == 1, "a attached once")
expect(a!.countSeenOnAttach == 0, "a saw an empty bus on attach (got \(a!.countSeenOnAttach))")
expect(a!.weakAttachedBus == nil, "bus wrapper handed to a was released after onAttached")
expect(bus!.subscriberCount() == 1, "bus still alive after callback released its reference")

expect(bus!.subscribe(subscriber: b!) == 2, "second subscribe returns 2")
expect(b!.attachedCount == 1, "b attached once")
expect(b!.countSeenOnAttach == 1, "b saw one subscriber on attach (got \(b!.countSeenOnAttach))")
expect(b!.attachedBus != nil && b!.weakAttachedBus != nil, "b kept the adopted bus")
expect(bus!.subscriberCount() == 2, "subscriberCount == 2")

// --- publish: Delivery steers the accepted count -----------------------------
expect(bus!.publish(topic: "news", text: "hello", tags: ["x", "y"]) == 2,
       "both subscribers accept 'news'")
expect(a!.routed == ["news"] && b!.routed == ["news"], "route asked once per subscriber")
expect(a!.received.count == 1 && b!.received.count == 1, "one delivery each")
let first = a!.received[0]
expect(first.seq == 1, "seq starts at 1 (got \(first.seq))")
expect(first.topic == "news", "topic (got \(first.topic))")
expect(first.text == "hello", "text (got \(first.text))")
expect(first.tags == ["x", "y"], "tags (got \(first.tags))")

// a skips 'quiet'; b accepts it.
expect(bus!.publish(topic: "quiet", text: "psst", tags: []) == 1, "skip lowers the accepted count")
expect(a!.routed == ["news", "quiet"], "a was asked about 'quiet' (got \(a!.routed))")
expect(a!.received.count == 1, "a did not receive the skipped message")
expect(b!.received.count == 2 && b!.received[1].seq == 2 && b!.received[1].tags.isEmpty,
       "b received the second message with empty tags")

// AcceptAndStop from the first subscriber stops delivery to the rest.
let c = TestSubscriber(name: "c")
expect(bus!.subscribe(subscriber: c) == 3, "third subscribe returns 3")
expect(bus!.publish(topic: "stop", text: "last", tags: ["z"]) == 1, "acceptAndStop delivers once")
expect(a!.received.count == 2 && a!.received[1].text == "last", "a took the stop message")
expect(b!.routed == ["news", "quiet"], "b was not asked after the stop (got \(b!.routed))")
expect(c.routed.isEmpty && c.received.isEmpty, "c was never reached")

// --- messages(): iterator as a lazy Sequence ---------------------------------
let texts = Array(bus!.messages())
expect(texts == ["hello", "psst", "last"], "messages in order (got \(texts))")
var joined: [String] = []
for m in bus!.messages() { joined.append(m.uppercased()) }
expect(joined == ["HELLO", "PSST", "LAST"], "for-in over the sequence")
expect(bus!.messages().map { $0.count } == [5, 4, 4], "map over the sequence")
// Abandoning early releases the producer iterator through deinit.
expect(bus!.messages().first(where: { $0.hasPrefix("p") }) == "psst", "early exit")

// --- lastMessage(): optional record ------------------------------------------
let last = bus!.lastMessage()
expect(last != nil, "lastMessage present")
expect(last!.seq == 3 && last!.topic == "stop" && last!.text == "last" && last!.tags == ["z"],
       "lastMessage fields (got \(last!))")

// --- publishLater: async wrapper awaited from the producer's thread ---------
let later = await bus!.publishLater(topic: "news", text: "later")
expect(later == 3, "publishLater accepted by all three (got \(later))")
expect(bus!.lastMessage()?.text == "later", "async publish logged")
expect(bus!.lastMessage()?.seq == 4, "async publish took seq 4")
expect(c.received.count == 1 && c.received[0].tags.isEmpty, "c received the async message")

// --- The bus adopted in onAttached is the same object --------------------------
let kept = b!.attachedBus!
expect(kept.subscriberCount() == 3, "kept bus sees all three subscribers")
expect(kept.publish(topic: "via-kept", text: "same object", tags: []) == 3,
       "publish through the kept wrapper reaches every subscriber")
expect(bus!.lastMessage()?.topic == "via-kept", "publish via kept wrapper visible through the original")
expect(Array(bus!.messages()).count == 5, "five messages logged")

// --- routeOnce: a callback passed to a free function is freed on return -------
weak var weakD: TestSubscriber?
do {
    let d = TestSubscriber(name: "d", skipTopic: "quiet")
    weakD = d
    expect(Events.routeOnce(subscriber: d, topic: "quiet") == .skip, "routeOnce skip")
    expect(Events.routeOnce(subscriber: d, topic: "stop") == .acceptAndStop, "routeOnce acceptAndStop")
    expect(Events.routeOnce(subscriber: d, topic: "other") == .accept, "routeOnce accept")
    expect(d.routed == ["quiet", "stop", "other"], "routeOnce asked route each time")
    expect(d.attachedCount == 0, "routeOnce never attaches")
}
expect(weakD == nil, "routeOnce released its subscriber (free ran)")

// --- clearSubscribers releases every subscriber ------------------------------
a = nil
b = nil
expect(weakA != nil && weakB != nil, "the bus retains a and b after the consumer drops them")
bus!.clearSubscribers()
expect(bus!.subscriberCount() == 0, "no subscribers after clear")
expect(weakA == nil, "a freed by clearSubscribers")
expect(weakB == nil, "b (and the bus reference it kept) freed by clearSubscribers")
expect(bus!.lastMessage()?.topic == "via-kept", "bus still alive after b's kept reference was released")
expect(bus!.publish(topic: "news", text: "nobody", tags: []) == 0, "publish with no subscribers")
expect(c.received.count == 2, "c is no longer delivered to (got \(c.received.count))")

// --- Destroying the bus frees a retained subscriber --------------------------
weak var weakE: TestSubscriber?
weak var weakBus2: EventBus?
do {
    let bus2 = EventBus()
    weakBus2 = bus2
    let e = TestSubscriber(name: "e")
    weakE = e
    expect(bus2.subscribe(subscriber: e) == 1, "bus2 subscribe")
    expect(bus2.publish(topic: "t", text: "m", tags: []) == 1, "bus2 publish")
}
expect(weakBus2 == nil, "bus2 wrapper deinitialized")
expect(weakE == nil, "dropping the last bus reference freed its subscriber")

bus = nil
expect(TestSubscriber.live == 1, "only c is still alive (got \(TestSubscriber.live))")

print("swift/events: OK")
